lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/ImpersonationRule.java

```diff
@@ -31,6 +32,7 @@
     private final Optional<Pattern> originalRolePattern;
     private final Pattern newUserPattern;
     private final boolean allow;
+    private final StringBuilder stringBuilder;
```

Comment:
StringBuilder is mutable object. Please replace it with something immutable here like `Pattern`

Also I think it should be a local variable in `match` method.

Reply:
Java 17 `Matcher` lacks a method `String substituteReplacement(String replacement)` which, unlike 
`replaceAll()` that replaces all matched groups in original string with substituted replacement, just simply returns the substituted replacement string, without replacing the original string.

Therefore, the only other viable option is to use `appendReplacement()`. However, this method either requires `StringBuffer` or `StringBuilder`, both mutable. In my earlier version, I had declared the `StringBuilder` object in `match` method, the only downside is that each method invocation will require a new `StringBuilder` to be constructed. It was lifted here for perf reasons. I will move it back to `match`.

Reply:
Done.